File: test/core/testability/testability_spec.dart
library angular2.test.core.testability.testability_spec;

import "package:angular2/testing_internal.dart"
    show
        AsyncTestCompleter,
        inject,
        describe,
        ddescribe,
        it,
        iit,
        xit,
        xdescribe,
        expect,
        beforeEach,
        SpyObject;
import "package:angular2/src/core/testability/testability.dart"
    show Testability;
import "package:angular2/src/core/zone/ng_zone.dart" show NgZone;
import "package:angular2/src/facade/async.dart"
    show PromiseWrapper, EventEmitter, ObservableWrapper;

// Schedules a microtasks (using a resolved promise .then())
void microTask(Function fn) {
  PromiseWrapper.resolve(null).then((_) {
    fn();
  });
}

class MockNgZone extends NgZone {
  EventEmitter<dynamic> _onTurnStartStream;
  get onTurnStart {
    return this._onTurnStartStream;
  }

  EventEmitter<dynamic> _onEventDoneStream;
  get onEventDone {
    return this._onEventDoneStream;
  }

  MockNgZone() : super(enableLongStackTrace: false) {
    /* super call moved to initializer */;
    this._onTurnStartStream = new EventEmitter(false);
    this._onEventDoneStream = new EventEmitter(false);
  }
  void start() {
    ObservableWrapper.callEmit(this._onTurnStartStream, null);
  }

  void finish() {
    ObservableWrapper.callEmit(this._onEventDoneStream, null);
  }
}

main() {
  describe("Testability", () {
    var testability, execute, ngZone;
    beforeEach(() {
      ngZone = new MockNgZone();
      testability = new Testability(ngZone);
      execute = new SpyObject().spy("execute");
    });
    describe("Pending count logic", () {
      it("should start with a pending count of 0", () {
        expect(testability.getPendingRequestCount()).toEqual(0);
      });
      it(
          "should fire whenstable callbacks if pending count is 0",
          inject([AsyncTestCompleter], (async) {
            testability.whenStable(execute);
            microTask(() {
              expect(execute).toHaveBeenCalled();
              async.done();
            });
          }));
      it("should not fire whenstable callbacks synchronously if pending count is 0",
          () {
        testability.whenStable(execute);
        expect(execute).not.toHaveBeenCalled();
      });
      it(
          "should not call whenstable callbacks when there are pending counts",
          inject([AsyncTestCompleter], (async) {
            testability.increasePendingRequestCount();
            testability.increasePendingRequestCount();
            testability.whenStable(execute);
            microTask(() {
              expect(execute).not.toHaveBeenCalled();
              testability.decreasePendingRequestCount();
              microTask(() {
                expect(execute).not.toHaveBeenCalled();
                async.done();
              });
            });
          }));
      it(
          "should fire whenstable callbacks when pending drops to 0",
          inject([AsyncTestCompleter], (async) {
            testability.increasePendingRequestCount();
            testability.whenStable(execute);
            microTask(() {
              expect(execute).not.toHaveBeenCalled();
              testability.decreasePendingRequestCount();
              microTask(() {
                expect(execute).toHaveBeenCalled();
                async.done();
              });
            });
          }));
      it("should not fire whenstable callbacks synchronously when pending drops to 0",
          () {
        testability.increasePendingRequestCount();
        testability.whenStable(execute);
        testability.decreasePendingRequestCount();
        expect(execute).not.toHaveBeenCalled();
      });
    });
    describe("NgZone callback logic", () {
      it("should start being ready", () {
        expect(testability.isAngularEventPending()).toEqual(false);
      });
      it(
          "should fire whenstable callback if event is already finished",
          inject([AsyncTestCompleter], (async) {
            ngZone.start();
            ngZone.finish();
            testability.whenStable(execute);
            microTask(() {
              expect(execute).toHaveBeenCalled();
              async.done();
            });
          }));
      it("should not fire whenstable callbacks synchronously if event is already finished",
          () {
        ngZone.start();
        ngZone.finish();
        testability.whenStable(execute);
        expect(execute).not.toHaveBeenCalled();
      });
      it(
          "should fire whenstable callback when event finishes",
          inject([AsyncTestCompleter], (async) {
            ngZone.start();
            testability.whenStable(execute);
            microTask(() {
              expect(execute).not.toHaveBeenCalled();
              ngZone.finish();
              microTask(() {
                expect(execute).toHaveBeenCalled();
                async.done();
              });
            });
          }));
      it("should not fire whenstable callbacks synchronously when event finishes",
          () {
        ngZone.start();
        testability.whenStable(execute);
        ngZone.finish();
        expect(execute).not.toHaveBeenCalled();
      });
      it(
          "should not fire whenstable callback when event did not finish",
          inject([AsyncTestCompleter], (async) {
            ngZone.start();
            testability.increasePendingRequestCount();
            testability.whenStable(execute);
            microTask(() {
              expect(execute).not.toHaveBeenCalled();
              testability.decreasePendingRequestCount();
              microTask(() {
                expect(execute).not.toHaveBeenCalled();
                ngZone.finish();
                microTask(() {
                  expect(execute).toHaveBeenCalled();
                  async.done();
                });
              });
            });
          }));
      it(
          "should not fire whenstable callback when there are pending counts",
          inject([AsyncTestCompleter], (async) {
            ngZone.start();
            testability.increasePendingRequestCount();
            testability.increasePendingRequestCount();
            testability.whenStable(execute);
            microTask(() {
              expect(execute).not.toHaveBeenCalled();
              ngZone.finish();
              microTask(() {
                expect(execute).not.toHaveBeenCalled();
                testability.decreasePendingRequestCount();
                microTask(() {
                  expect(execute).not.toHaveBeenCalled();
                  testability.decreasePendingRequestCount();
                  microTask(() {
                    expect(execute).toHaveBeenCalled();
                    async.done();
                  });
                });
              });
            });
          }));
    });
  });
}

File: packages/core/test/testability/testability_spec.ts
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.dev/license
 */

import {EventEmitter} from '../../src/core';
import {Injectable} from '../../src/di';
import {
  GetTestability,
  PendingMacrotask,
  Testability,
  TestabilityRegistry,
  setTestabilityGetter,
} from '../../src/testability/testability';
import {NgZone} from '../../src/zone/ng_zone';
import {fakeAsync, tick, waitForAsync} from '../../testing';

// Schedules a microtasks (using queueMicrotask)
function microTask(fn: Function): void {
  queueMicrotask(() => {
    // We do double dispatch so that we can wait for queueMicrotask in the Testability when
    // NgZone becomes stable.
    queueMicrotask(() => fn());
  });
}

class NoopGetTestability implements GetTestability {
  addToWindow(registry: TestabilityRegistry): void {}
  findTestabilityInTree(
    registry: TestabilityRegistry,
    elem: any,
    findInAncestors: boolean,
  ): Testability | null {
    return null;
  }
}

@Injectable()
class MockNgZone extends NgZone {
  /** @internal */
  override onUnstable: EventEmitter<any>;

  /** @internal */
  override onStable: EventEmitter<any>;

  constructor() {
    super({enableLongStackTrace: false});
    this.onUnstable = new EventEmitter(false);
    this.onStable = new EventEmitter(false);
  }

  unstable(): void {
    this.onUnstable.emit(null);
  }

  stable(): void {
    this.onStable.emit(null);
  }
}

describe('Testability', () => {
  let testability: Testability;
  let execute: any;
  let execute2: any;
  let updateCallback: any;
  let ngZone: MockNgZone;

  beforeEach(waitForAsync(() => {
    ngZone = new MockNgZone();
    testability = new Testability(ngZone, new TestabilityRegistry(), new NoopGetTestability());
    execute = jasmine.createSpy('execute');
    execute2 = jasmine.createSpy('execute');
    updateCallback = jasmine.createSpy('execute');
  }));
  afterEach(() => {
    // Instantiating the Testability (via `new Testability` above) has a side
    // effect of defining the testability getter globally to a specified value.
    // This call resets that reference after each test to make sure it does not
    // get leaked between tests. In real scenarios this is not a problem, since
    // the `Testability` is created via DI and uses the same testability getter
    // (injected into a constructor) across all instances.
    setTestabilityGetter(null! as GetTestability);
  });

  describe('NgZone callback logic', () => {
    describe('whenStable with timeout', () => {
      it('should fire if Angular is already stable', waitForAsync(() => {
        testability.whenStable(execute, 200);

        microTask(() => {
          expect(execute).toHaveBeenCalled();
        });
      }));

      it('should fire when macroTasks are cancelled', fakeAsync(() => {
        const id = ngZone.run(() => setTimeout(() => {}, 1000));
        testability.whenStable(execute, 500);

        tick(200);
        ngZone.run(() => clearTimeout(id));
        // fakeAsync doesn't trigger NgZones whenStable
        ngZone.stable();

        tick(1);
        expect(execute).toHaveBeenCalled();
      }));

      it('calls the done callback when angular is stable', fakeAsync(() => {
        let timeout1Done = false;
        ngZone.run(() => setTimeout(() => (timeout1Done = true), 500));
        testability.whenStable(execute, 1000);

        tick(600);
        ngZone.stable();
        tick();

        expect(timeout1Done).toEqual(true);
        expect(execute).toHaveBeenCalled();

        // Should cancel the done timeout.
        tick(500);
        ngZone.stable();
        tick();
        expect(execute.calls.count()).toEqual(1);
      }));

      it('calls update when macro tasks change', fakeAsync(() => {
        let timeout1Done = false;
        let timeout2Done = false;
        ngZone.run(() => setTimeout(() => (timeout1Done = true), 500));
        tick();
        testability.whenStable(execute, 1000, updateCallback);

        tick(100);
        ngZone.run(() => setTimeout(() => (timeout2Done = true), 300));
        expect(updateCallback.calls.count()).toEqual(1);
        tick(600);

        expect(timeout1Done).toEqual(true);
        expect(timeout2Done).toEqual(true);
        expect(updateCallback.calls.count()).toEqual(3);
        expect(execute).toHaveBeenCalled();

        const update1 = updateCallback.calls.all()[0].args[0] as PendingMacrotask[];
        expect(update1[0].data!.delay).toEqual(500);

        const update2 = updateCallback.calls.all()[1].args[0] as PendingMacrotask[];
        expect(update2[0].data!.delay).toEqual(500);
        expect(update2[1].data!.delay).toEqual(300);
      }));

      it('cancels the done callback if the update callback returns true', fakeAsync(() => {
        let timeoutDone = false;
        ngZone.unstable();
        execute2.and.returnValue(true);
        testability.whenStable(execute, 1000, execute2);

        tick(100);
        ngZone.run(() => setTimeout(() => (timeoutDone = true), 500));
        ngZone.stable();
        expect(execute2).toHaveBeenCalled();

        tick(500);
        ngZone.stable();
        tick();

        expect(execute).not.toHaveBeenCalled();
      }));
    });

    it('should fire whenstable callback if event is already finished', fakeAsync(() => {
      ngZone.unstable();
      ngZone.stable();
      testability.whenStable(execute);

      tick();
      expect(execute).toHaveBeenCalled();
    }));

    it('should not fire whenstable callbacks synchronously if event is already finished', () => {
      ngZone.unstable();
      ngZone.stable();
      testability.whenStable(execute);

      expect(execute).not.toHaveBeenCalled();
    });

    it('should fire whenstable callback when event finishes', fakeAsync(() => {
      ngZone.unstable();
      testability.whenStable(execute);

      tick();
      expect(execute).not.toHaveBeenCalled();
      ngZone.stable();

      tick();
      expect(execute).toHaveBeenCalled();
    }));

    it('should not fire whenstable callbacks synchronously when event finishes', () => {
      ngZone.unstable();
      testability.whenStable(execute);
      ngZone.stable();

      expect(execute).not.toHaveBeenCalled();
    });

    it('should fire whenstable callback with didWork if event is already finished', fakeAsync(() => {
      ngZone.unstable();
      ngZone.stable();
      testability.whenStable(execute);

      tick();
      expect(execute).toHaveBeenCalled();
      testability.whenStable(execute2);

      tick();
      expect(execute2).toHaveBeenCalled();
    }));

    it('should fire whenstable callback with didwork when event finishes', fakeAsync(() => {
      ngZone.unstable();
      testability.whenStable(execute);

      tick();
      ngZone.stable();

      tick();
      expect(execute).toHaveBeenCalled();
      testability.whenStable(execute2);

      tick();
      expect(execute2).toHaveBeenCalled();
    }));
  });
});

describe('TestabilityRegistry', () => {
  let testability1: Testability;
  let testability2: Testability;
  let registry: TestabilityRegistry;
  let ngZone: MockNgZone;

  beforeEach(waitForAsync(() => {
    ngZone = new MockNgZone();
    registry = new TestabilityRegistry();
    testability1 = new Testability(ngZone, registry, new NoopGetTestability());
    testability2 = new Testability(ngZone, registry, new NoopGetTestability());
  }));
  afterEach(() => {
    // Instantiating the Testability (via `new Testability` above) has a side
    // effect of defining the testability getter globally to a specified value.
    // This call resets that reference after each test to make sure it does not
    // get leaked between tests. In real scenarios this is not a problem, since
    // the `Testability` is created via DI and uses the same testability getter
    // (injected into a constructor) across all instances.
    setTestabilityGetter(null! as GetTestability);
  });
  describe('unregister testability', () => {
    it('should remove the testability when unregistering an existing testability', () => {
      registry.registerApplication('testability1', testability1);
      registry.registerApplication('testability2', testability2);
      registry.unregisterApplication('testability2');
      expect(registry.getAllTestabilities().length).toEqual(1);
      expect(registry.getTestability('testability1')).toEqual(testability1);
    });

    it('should remain the same when unregistering a non-existing testability', () => {
      expect(registry.getAllTestabilities().length).toEqual(0);
      registry.registerApplication('testability1', testability1);
      registry.registerApplication('testability2', testability2);
      registry.unregisterApplication('testability3');
      expect(registry.getAllTestabilities().length).toEqual(2);
      expect(registry.getTestability('testability1')).toEqual(testability1);
      expect(registry.getTestability('testability2')).toEqual(testability2);
    });

    it('should remove all the testability when unregistering all testabilities', () => {
      registry.registerApplication('testability1', testability1);
      registry.registerApplication('testability2', testability2);
      registry.unregisterAllApplications();
      expect(registry.getAllTestabilities().length).toEqual(0);
    });
  });
});
